server/world/recipes/item.go

```diff
@@ -0,0 +1,114 @@
+package recipes
+
+import (
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+)
+
+// Item is a recipe item. It is an item stack inherently, but it also has an extra value for if it applies to all types.
+type Item struct {
+	item.Stack
+	// AllTypes is true if if the item applies to all of it's type. This is so it can be rendered in the recipe book
+	// for alternative types.
+	AllTypes bool
+}
+
+// ItemType is an item that has a name and a metadata value.
+type ItemType struct {
+	// Name is the name of the item type.
+	Name string `nbt:"name"`
+	// MetadataValue is the meta of the item type.
+	MetadataValue int32 `nbt:"meta"`
+}
+
+// ToItem converts the item type to an item.
+func (i ItemType) ToItem() (it world.Item, ok bool) {
+	return world.ItemByName(i.Name, int16(i.MetadataValue))
+}
+
+// InputItem is an item that is inputted to a crafting menu.
+type InputItem struct {
+	// Name is the name of the item being inputted.
+	Name string `nbt:"name"`
+	// MetadataValue is the meta of the item. This can change the item almost completely, or act as durability.
+	MetadataValue int32 `nbt:"meta"`
+	// Count is the amount of the item.
+	Count int32 `nbt:"count"`
+}
```

Comment:
This is rather awful to expose to the API.

Reply:
Unexported.